File: bank-service/src/main/java/rs/edu/raf/security/SpringSecurityConfig.java
package rs.edu.raf.security;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.builders.WebSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityCustomizer;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

import java.util.Arrays;
import java.util.List;

@Configuration
@EnableWebSecurity
public class SpringSecurityConfig implements WebSecurityCustomizer {


    private final JwtFilter jwtFilter;

    @Autowired
    public SpringSecurityConfig(JwtFilter jwtFilter) {
        this.jwtFilter = jwtFilter;
    }

    @Bean
    public BCryptPasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    public AuthenticationManager authenticationManager(AuthenticationConfiguration config) throws Exception {
        return config.getAuthenticationManager();
    }

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity httpSecurity) throws Exception {
        httpSecurity
                .cors(cors -> cors.configurationSource(corsConfigurationSource()))
                .authorizeRequests(authorize -> authorize
                        .requestMatchers(HttpMethod.OPTIONS, "/**").permitAll()
                                .requestMatchers("/swagger-ui/**", "/api-docs/**", "/swagger-resources/**",
                                        "/exchange", "/bank-accounts/deduct-available-balance/**").permitAll()
                        .requestMatchers("/bank-accounts/client", "/cards/client", "/credit/requests/*").hasAuthority("ROLE_CLIENT")
                        .requestMatchers(HttpMethod.GET, "/country", "/currency",
                                "/card-issuer", "/credit/*", "/credit-type", "/bank-accounts/account/**", "/bank-accounts/account/number/**").authenticated()
                        .requestMatchers("/credit/apply").authenticated()
                        .requestMatchers("/bank-accounts", "/bank-accounts/**", "/country", "/country/**",
                                "/currency", "/currency/**").hasAuthority("ROLE_MANAGE_BANK_ACCOUNTS")
                        .requestMatchers("/cards", "/cards/**", "/card-issuer", "/card-issuer/**").hasAuthority("ROLE_MANAGE_CARDS")
                        .requestMatchers("/credit","/credit/**", "/credit-type", "/credit-type/**").hasAuthority("ROLE_MANAGE_CREDITS")
.anyRequest().authenticated()

                )
                .csrf().disable()
                .addFilterBefore(this.jwtFilter, UsernamePasswordAuthenticationFilter.class);
        return httpSecurity.build();
    }

    @Bean
    public CorsConfigurationSource corsConfigurationSource() {
        CorsConfiguration configuration = new CorsConfiguration();

        configuration.setAllowedOrigins(List.of("http://localhost:5002"));

        configuration.setAllowedMethods(Arrays.asList("GET", "POST", "PUT", "DELETE", "OPTIONS"));
        configuration.setAllowedHeaders(Arrays.asList("Authorization", "Cache-Control", "Content-Type"));
        configuration.setAllowCredentials(true);

        UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
        source.registerCorsConfiguration("/**", configuration);
        return source;
    }

    @Override
    public void customize(WebSecurity web) {
        web.ignoring().requestMatchers(new AntPathRequestMatcher("/h2-console/**"));
    }

}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/security/SpringSecurityConfig.kt
package org.example.bankservicekotlin.security

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpMethod
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.config.Customizer
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.builders.WebSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityCustomizer
import org.springframework.security.config.annotation.web.configurers.CorsConfigurer
import org.springframework.security.config.annotation.web.configurers.ExpressionUrlAuthorizationConfigurer.ExpressionInterceptUrlRegistry
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource

@Configuration
@EnableWebSecurity
class SpringSecurityConfig(val jwtFilter: JwtFilter):WebSecurityCustomizer {
    @Bean
    fun passwordEncoder(): BCryptPasswordEncoder {
        return BCryptPasswordEncoder()
    }

    @Bean
    @Throws(Exception::class)
    fun authenticationManager(config: AuthenticationConfiguration): AuthenticationManager {
        return config.authenticationManager
    }

    @Bean
    @Throws(Exception::class)
    fun securityFilterChain(httpSecurity: HttpSecurity): SecurityFilterChain {
        httpSecurity
            .cors { cors: CorsConfigurer<HttpSecurity?> -> cors.configurationSource(corsConfigurationSource()) }
            .authorizeHttpRequests {
                it
                    .requestMatchers(HttpMethod.OPTIONS, "/**").permitAll()
                    .requestMatchers(
                        "/swagger-ui/**", "/api-docs/**", "/swagger-resources/**",
                        "/exchange", "/bank-accounts/deduct-available-balance/**"
                    ).permitAll()
                    .requestMatchers(
                        "/bank-accounts/client",
                        "/cards/client",
                        "/credit/requests/*",
                        "/bank-accounts/client/**"
                    ).hasAuthority("ROLE_CLIENT")
                    .requestMatchers(
                        HttpMethod.GET,
                        "/country",
                        "/currency",
                        "/card-issuer",
                        "/credit/*",
                        "/credit-type",
                        "/bank-accounts/account/**",
                        "/bank-accounts/account/number/**"
                    ).authenticated()
                    .requestMatchers("/credit/apply").authenticated()
                    .requestMatchers(
                        "/bank-accounts", "/bank-accounts/**", "/country", "/country/**",
                        "/currency", "/currency/**"
                    ).hasAuthority("ROLE_MANAGE_BANK_ACCOUNTS")
                    .requestMatchers("/cards", "/cards/**", "/card-issuer", "/card-issuer/**")
                    .hasAuthority("ROLE_MANAGE_CARDS")
                    .requestMatchers("/credit", "/credit/**", "/credit-type", "/credit-type/**")
                    .hasAuthority("ROLE_MANAGE_CREDITS")
                    .anyRequest().authenticated()
            }
            .csrf { it.disable() }
            .addFilterBefore(this.jwtFilter, UsernamePasswordAuthenticationFilter::class.java)
        return httpSecurity.build()
    }

    @Bean
    fun corsConfigurationSource(): CorsConfigurationSource {
        val configuration = CorsConfiguration()

        configuration.allowedOrigins = listOf("http://localhost:5002")

        configuration.allowedMethods = mutableListOf("GET", "POST", "PUT", "DELETE", "OPTIONS")
        configuration.allowedHeaders =
            mutableListOf("Authorization", "Cache-Control", "Content-Type")
        configuration.allowCredentials = true

        val source = UrlBasedCorsConfigurationSource()
        source.registerCorsConfiguration("/**", configuration)
        return source
    }

    override fun customize(web: WebSecurity?) {

    }
}